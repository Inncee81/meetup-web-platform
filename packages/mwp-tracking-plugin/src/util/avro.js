// @flow
const log = require('mwp-logger-plugin').logger;
const avro = require('avsc');

/*
 * There are currently 2 distinct analytics logging methods
 * 1. `stdout`, which will be consumed automatically from apps running in
 *    Google Container Engine
 * 2. Google Pub/Sub, which is used from inside Google App Engine
 *
 * The Pub/Sub setup _only_ works when the environment is configured to
 * automatically authorize Pub/Sub messages, so we use an `isGAE` env variable
 * to determine whether the app is running in GAE.
 *
 * @see https://meetup.atlassian.net/wiki/display/MUP/Analytics+Logging#AnalyticsLogging-Theinputmechanism
 */
const getPlatformAnalyticsLog = (
	isGAE: ?string = process.env.GAE_INSTANCE
): (string => void) => {
	if (isGAE) {
		const pubsub = require('@google-cloud/pubsub')({
			projectId: 'meetup-prod',
		});
		const analyticsLog = pubsub.topic('analytics-log-json');
		return (serializedRecord: string) => {
			analyticsLog.publish(serializedRecord).then(
				(messageIds, apiResponse) => {
					if (messageIds) {
						log.info({ messageIds }, 'GAE PubSub');
					}
				},
				err => log.error(err)
			);
		};
	}
	return (serializedRecord: string) => {
		process.stdout.write(`analytics=${serializedRecord}\n`);
	};
};

const analyticsLog = getPlatformAnalyticsLog();
const debugLog = deserializedRecord =>
	console.log(JSON.stringify(deserializedRecord));

// currently the schema is manually copied from
// https://github.dev.meetup.com/meetup/meetup/blob/master/modules/base/src/main/versioned_avro/Click_v2.avsc
const click = {
	namespace: 'com.meetup.base.avro',
	type: 'record',
	name: 'Click',
	doc: 'v2',
	fields: [
		{ name: 'timestamp', type: 'string' },
		{ name: 'requestId', type: 'string' },
		{ name: 'memberId', type: 'int' },
		{ name: 'lineage', type: 'string' },
		{ name: 'linkText', type: 'string' },
		{ name: 'coordX', type: 'int' },
		{ name: 'coordY', type: 'int' },
		{ name: 'tag', type: 'string', default: '' },
	],
};

// currently the schema is manually copied from
// https://github.com/meetup/meetup/blob/master/modules/base/src/main/versioned_avro/Activity_v6.avsc
const activity = {
	namespace: 'com.meetup.base.avro',
	type: 'record',
	name: 'Activity',
	doc: 'v6',
	fields: [
		{ name: 'requestId', type: 'string' },
		{ name: 'timestamp', type: 'string' },
		{ name: 'url', type: 'string' },
		{ name: 'aggregratedUrl', type: 'string', default: '' }, // it's misspelled in the original spec
		{ name: 'ip', type: 'string', default: '' },
		{ name: 'agent', type: 'string', default: '' },
		{ name: 'memberId', type: 'int' },
		{ name: 'trackId', type: 'string' },
		{ name: 'mobileWeb', type: 'boolean' },
		{ name: 'platform', type: 'string' },
		{ name: 'referer', type: 'string' }, // it's misspelled in the original spec
		{ name: 'trax', type: { type: 'map', values: 'string' } },
		{
			name: 'platformAgent',
			type: {
				type: 'enum',
				name: 'PlatformAgent',
				symbols: [
					'WEB',
					'MUP_WEB',
					'PRO_WEB',
					'NATIVE',
					'NATIVE_APP_WEB_VIEW',
					'THIRD_PARTY_UNKNOWN',
					'UNKNOWN',
				],
			},
			default: 'UNKNOWN',
		},
		{ name: 'isUserActivity', type: 'boolean', default: true },
		{ name: 'browserId', type: 'string', default: '' },
	],
};

type Serializer = Object => string;
type Deserializer = string => Object;

const avroSerializer: Object => Serializer = schema => data => {
	const record = avro.parse(schema).toBuffer(data);
	// data.timestamp _must_ be ISOString if it exists
	const timestamp = data.timestamp || new Date().toISOString();
	const analytics = {
		record: record.toString('base64'),
		schema: `gs://meetup-logs/avro_schemas/${schema.name}_${schema.doc}.avsc`,
		date: timestamp.substr(0, 10), // YYYY-MM-DD
	};
	return JSON.stringify(analytics);
};

const avroDeserializer: Object => Deserializer = schema => serialized => {
	const { record } = JSON.parse(serialized);
	const avroBuffer = new Buffer(record, 'base64');
	return avro.parse(schema).fromBuffer(avroBuffer);
};

const logger = (serializer: Serializer, deserializer: Deserializer) => (
	record: Object
) => {
	const serializedRecord = serializer(record);
	const deserializedRecord = deserializer(serializedRecord);
	analyticsLog(serializedRecord);
	debugLog(deserializedRecord);
};

const schemas = {
	activity,
	click,
};
const serializers = {
	avro: avroSerializer,
	activity: avroSerializer(schemas.activity),
	click: avroSerializer(schemas.click),
};
const deserializers = {
	avro: avroDeserializer,
	activity: avroDeserializer(schemas.activity),
	click: avroDeserializer(schemas.click),
};
const loggers = {
	activity: logger(serializers.activity, deserializers.activity),
	click: logger(serializers.click, deserializers.click),
};

module.exports = {
	avroSerializer,
	getPlatformAnalyticsLog,
	schemas,
	serializers,
	deserializers,
	loggers,
};
