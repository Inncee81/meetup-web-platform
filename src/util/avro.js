const avro = require('avsc');

// currently the schema is manually copied from
// https://github.dev.meetup.com/meetup/meetup/blob/master/modules/base/src/main/versioned_avro/Click_v2.avsc
const click = {
	namespace: 'com.meetup.base.avro',
	type: 'record',
	name: 'Click',
	doc: 'v2',
	fields: [
		{ name: 'timestamp', type: 'string' },
		{ name: 'requestId', type: 'string' },
		{ name: 'memberId', type: 'int' },
		{ name: 'lineage', type: 'string' },
		{ name: 'linkText', type: 'string' },
		{ name: 'coordX', type: 'int' },
		{ name: 'coordY', type: 'int' },
		{ name: 'tag', type: 'string', default: '' },
	],
};

// currently the schema is manually copied from
// https://github.dev.meetup.com/meetup/meetup/blob/master/modules/base/src/main/versioned_avro/Activity_v3.avsc
const activity = {
	namespace: 'com.meetup.base.avro',
	type: 'record',
	name: 'Activity',
	doc: 'v3',
	fields: [
		{ name: 'requestId', type: 'string' },
		{ name: 'timestamp', type: 'string' },
		{ name: 'url', type: 'string' },
		{ name: 'aggregratedUrl', type: 'string', default: '' }, // it's misspelled in the original spec
		{ name: 'ip', type: 'string', default: '' },
		{ name: 'agent', type: 'string', default: '' },
		{ name: 'memberId', type: 'int' },
		{ name: 'trackId', type: 'string' },
		{ name: 'mobileWeb', type: 'boolean' },
		{ name: 'platform', type: 'string' },
		{ name: 'referer', type: 'string' }, // it's misspelled in the original spec
		{ name: 'trax', type: { type: 'map', values: 'string' } },
		{
			name: 'platformAgent',
			type: {
				type: 'enum',
				name: 'PlatformAgent',
				symbols: [
					'WEB',
					'NATIVE',
					'NATIVE_APP_WEB_VIEW',
					'THIRD_PARTY_UNKNOWN',
					'UNKNOWN',
				],
			},
			default: 'UNKNOWN',
		},
	],
};

/**
 * @param {Object} schema an avro JSON schema (.avsc)
 * @param {Object} data the data matching the schema.
 * @return {String} JSON string of the avro-encoded record wrapped with metadata
 */
const avroSerializer = schema => data => {
	const record = avro.parse(schema).toBuffer(data);
	// data.timestamp _must_ be ISOString if it exists
	const timestamp = data.timestamp || new Date().toISOString();
	const analytics = {
		record: record.toString('base64'),
		schema: `gs://meetup-logs/avro_schemas/${schema.name}_${schema.doc}.avsc`,
		date: timestamp.substr(0, 10), // YYYY-MM-DD
	};
	return `analytics=${JSON.stringify(analytics)}\n`;
};

const logger = serializer => record => {
	const serializedRecord = serializer(record);
	process.stdout.write(serializedRecord);
};

const schemas = {
	activity,
	click,
};
const serializers = {
	avro: avroSerializer,
	activity: avroSerializer(schemas.activity),
	click: avroSerializer(schemas.click),
};
const loggers = {
	activity: logger(serializers.activity),
	click: logger(serializers.click),
};

module.exports = {
	avroSerializer,
	serializers,
	schemas,
	loggers,
};
