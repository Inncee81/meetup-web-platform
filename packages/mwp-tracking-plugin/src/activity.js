// @flow
import avro from './util/avro';
import { getTrackActivity, getTrackApiResponses } from './_activityTrackers';
import { ACTIVITY_PLUGIN_NAME } from './config';
import { getISOStringNow } from './util/trackingUtils';

const YEAR_IN_MS: number = 1000 * 60 * 60 * 24 * 365;

/*
 * This plugin provides `request.track...` methods that track events related to
 * particular server responses, e.g. new sesssions and navigation activity.
 *
 * Available trackers:
 * - `trackActivity`
 */

type ActivityPlatform = 'WEB' | 'IOS' | 'ANDROID';
const ANDROID_APP_ID = 'com.meetup';

export const getRequestPlatform = (request: HapiRequest): ActivityPlatform => {
	const { headers, state, query = {} } = request;
	const isNativeApp = state.isNativeApp || query.isNativeApp;
	if (isNativeApp) {
		// recommended test for Android WebView - not perfect but should be adequate
		// https://stackoverflow.com/questions/24291315/android-webview-detection-in-php
		const isAndroid = headers.http_x_requested_with === ANDROID_APP_ID;
		return isAndroid ? 'ANDROID' : 'IOS';
	}
	return 'WEB';
};

export const getLogger: string => (Object, Object) => mixed = (agent: string) => {
	return (request: Object, trackInfo: Object) => {
		const { headers } = request;

		const record = {
			timestamp: getISOStringNow(),
			requestId: request.id,
			ip: headers['remote-addr'] || '',
			agent: headers['user-agent'] || '',
			platform: getRequestPlatform(request),
			platformAgent: agent,
			mobileWeb: false,
			referer: '', // misspelled to align with schema
			trax: {},
			isUserActivity: true,
			...trackInfo,
		};

		avro.loggers.awsactivity(record);
		return record;
	};
};

/*
 * Each tracking function is a composition of a logging function and data about
 * the `request` object. This function computes some configuration information
 * to create the tracking functions, and returns each of them in a map keyed by
 * the target `request` method name
 */
export function getTrackers(options: {
	agent: string,
	browserIdCookieName: string,
	memberCookieName: string,
	trackIdCookieName: string,
}): { [string]: Tracker } {
	const {
		agent,
		browserIdCookieName,
		memberCookieName,
		trackIdCookieName,
	} = options;

	const log: Logger = getLogger(agent);
	const trackOpts: TrackOpts = {
		log,
		browserIdCookieName,
		memberCookieName,
		trackIdCookieName,
	};
	// These are the tracking methods that will be set on the `request` interface
	const trackers: { [string]: Tracker } = {
		trackActivity: getTrackActivity(trackOpts),
		trackApiResponses: getTrackApiResponses(trackOpts),
	};
	return trackers;
}

/*
 * Run request-initialization routines, e.g. creating plugin data store
 */
const onRequest = (request, h) => {
	// initialize request.plugins[ACTIVITY_PLUGIN_NAME] to store cookie vals
	request.plugins[ACTIVITY_PLUGIN_NAME] = {};
	return h.continue;
};
/*
 * Read from request data to prepare/modify response. Mainly looking for new
 * tracking cookies that need to be set using request.response.state.
 */

type CookieOpts = {
	browserIdCookieName: string,
	memberCookieName: string,
	trackIdCookieName: string,
	domain: string,
};
export const getOnPreResponse = (cookieConfig: CookieOpts) => (
	request: HapiRequest,
	h: HapiResponseToolkit
) => {
	const { browserIdCookieName, trackIdCookieName, domain } = cookieConfig;
	const pluginData = request.plugins[ACTIVITY_PLUGIN_NAME];
	const browserId = pluginData[browserIdCookieName];
	const trackId = pluginData[trackIdCookieName];

	const FOREVER: HapiServerStateCookieOptions = {
		domain,
		encoding: 'none',
		path: '/',
		isHttpOnly: true,
		ttl: YEAR_IN_MS * 20,
		strictHeader: false, // skip strict cookie format validation (no quotes)
	};

	if (!request.response.isBoom) {
		if (browserId) {
			h.state(browserIdCookieName, browserId, FOREVER);
		}
		if (trackId) {
			h.state(trackIdCookieName, trackId, FOREVER);
		}
	}

	return h.continue;
};

/*
 * The plugin register function that will 'decorate' the `request` interface with
 * all tracking functions returned from `getTrackers`, as well as assign request
 * lifecycle event handlers that can affect the response, e.g. by setting cookies
 */
export function register(
	server: Object,
	options: { agent: string, isProdApi: boolean }
) {
	const { agent, isProdApi } = options;

	const trackIdCookieName: string = isProdApi ? 'MEETUP_TRACK' : 'MEETUP_TRACK_DEV';
	const browserIdCookieName: string = isProdApi
		? 'MEETUP_BROWSER_ID'
		: 'MEETUP_BROWSER_ID_DEV';
	const memberCookieName: string = isProdApi
		? 'MEETUP_MEMBER'
		: 'MEETUP_MEMBER_DEV';

	const trackers: { [string]: Tracker } = getTrackers({
		agent,
		browserIdCookieName,
		memberCookieName,
		trackIdCookieName,
	});

	Object.keys(trackers).forEach((trackType: string) => {
		server.decorate('request', trackType, trackers[trackType], {
			apply: true, // make the `request` available to tracker
		});
	});

	server.ext('onRequest', onRequest);
	server.ext(
		'onPreResponse',
		getOnPreResponse({
			browserIdCookieName,
			memberCookieName,
			trackIdCookieName,
			domain: isProdApi ? '.meetup.com' : '.dev.meetup.com',
		})
	);
}

export const plugin = {
	register,
	name: ACTIVITY_PLUGIN_NAME,
	version: '1.0.0',
};
