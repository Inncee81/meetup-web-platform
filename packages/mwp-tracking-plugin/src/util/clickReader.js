import { parseIdCookie } from './idUtils'; // TODO: provide this info through new plugin
import avro from './avro';
import { COOKIE_NAME } from './clickState';

/*
 * This module provides utilities for reading click cookie data sent as part of
 * a request to the application - it should be set up as a server plugin
 */

const isProd = process.env.NODE_ENV === 'production';
const memberCookieName = isProd ? 'MEETUP_MEMBER' : 'MEETUP_MEMBER_DEV';
export const clickCookieOptions = {
	isSecure: isProd,
	isHttpOnly: false,
	domain: `${isProd ? '' : '.dev'}.meetup.com`,
};

export const clickToClickRecord = request => click => {
	return {
		timestamp: click.timestamp || new Date().toISOString(),
		requestId: request.id,
		memberId: parseIdCookie(
			request.state[memberCookieName].toString() || '',
			true
		), // force integer
		lineage: click.lineage,
		linkText: click.linkText || '',
		coordX: click.coords[0] || 0,
		coordY: click.coords[1] || 0,
	};
};

export default function processClickTracking(request, reply) {
	const rawCookieValue = (request.state || {})[COOKIE_NAME];
	// It's possible that multiple cookies with the same value were sent, e.g.
	// one value for .dev.meetup.com and another for .meetup.com - parse only the first
	const cookieValue =
		rawCookieValue instanceof Array ? rawCookieValue[0] : rawCookieValue;
	if (!cookieValue || cookieValue === 'undefined') {
		return;
	}

	const cookieJSON = decodeURIComponent(cookieValue);
	const { history } = JSON.parse(cookieJSON);
	history.map(clickToClickRecord(request)).forEach(avro.loggers.click);

	reply.unstate(COOKIE_NAME, clickCookieOptions);
	return;
}
