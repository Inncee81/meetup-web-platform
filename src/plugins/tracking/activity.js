// @flow
import avro from './util/avro';
import { getTrackActivity, getTrackApiResponses } from './_activityTrackers';

const YEAR_IN_MS: number = 1000 * 60 * 60 * 24 * 365;
export const ACTIVITY_PLUGIN_NAME = 'tracking';

/*
 * This plugin provides `request.track...` methods that track events related to
 * particular server responses, e.g. new sesssions and navigation activity.
 *
 * Available trackers:
 * - `trackActivity`
 */

export const getLogger: string => (Object, Object) => mixed = (
	agent: string
) => (request: Object, trackInfo: Object) => {
	const requestHeaders = request.headers;
	
	// Takes in desired time to convert and applies offset
	const offset: String = new Date().getTimezoneOffset() * 60000; // gets detected timezone + converts to milliseconds
	const nowUTC: Date = new Date(Date.now() + offset);

	// generates new date object taking the time in milliseconds and
	// adds the runtime environment's timezone offset
	const NY_OFFSET: String = -14400000;
	const newNow: Date = new Date(nowUTC.getTime() - NY_OFFSET);

	const record = {
		timestamp: newNow.toISOString(),
		requestId: request.id,
		ip: requestHeaders['remote-addr'] || '',
		agent: requestHeaders['user-agent'] || '',
		platform: 'WEB',
		platformAgent: agent,
		mobileWeb: false,
		referer: '', // misspelled to align with schema
		trax: {},
		isUserActivity: true,
		...trackInfo,
	};

	avro.loggers.activity(record);
	return record;
};

/*
 * Each tracking function is a composition of a logging function and data about
 * the `request` object. This function computes some configuration information
 * to create the tracking functions, and returns each of them in a map keyed by
 * the target `request` method name
 */
export function getTrackers(options: {
	agent: string,
	browserIdCookieName: string,
	memberCookieName: string,
	trackIdCookieName: string,
}): { [string]: Tracker } {
	const {
		agent,
		browserIdCookieName,
		memberCookieName,
		trackIdCookieName,
	} = options;

	const log: Logger = getLogger(agent);
	const trackOpts: TrackOpts = {
		log,
		browserIdCookieName,
		memberCookieName,
		trackIdCookieName,
	};
	// These are the tracking methods that will be set on the `request` interface
	const trackers: { [string]: Tracker } = {
		trackActivity: getTrackActivity(trackOpts),
		trackApiResponses: getTrackApiResponses(trackOpts),
	};
	return trackers;
}

/*
 * Run request-initialization routines, e.g. creating plugin data store
 */
const onRequest = (request, reply) => {
	// initialize request.plugins[ACTIVITY_PLUGIN_NAME] to store cookie vals
	request.plugins[ACTIVITY_PLUGIN_NAME] = {};
	reply.continue();
};
/*
 * Read from request data to prepare/modify response. Mainly looking for new
 * tracking cookies that need to be set using request.response.state.
 */
const getOnPreResponse = cookieConfig => (request, reply) => {
	const { browserIdCookieName, trackIdCookieName } = cookieConfig;
	const pluginData = request.plugins[ACTIVITY_PLUGIN_NAME];
	const browserId = pluginData.browserIdCookieName;
	const trackId = pluginData.trackIdCookieName;

	const FOREVER: CookieOpts = {
		encoding: 'none',
		path: '/',
		isHttpOnly: true,
		ttl: YEAR_IN_MS * 20,
	};

	if (browserId) {
		request.response.state(browserIdCookieName, browserId, FOREVER);
	}
	if (trackId) {
		request.response.state(trackIdCookieName, trackId, FOREVER);
	}
	reply.continue();
};

/*
 * The plugin register function that will 'decorate' the `request` interface with
 * all tracking functions returned from `getTrackers`, as well as assign request
 * lifecycle event handlers that can affect the response, e.g. by setting cookies
 */
export default function register(
	server: Object,
	options: { agent: string, isProd: boolean },
	next: () => void
) {
	const { agent, isProd } = options;

	const trackIdCookieName: string = isProd
		? 'MEETUP_TRACK'
		: 'MEETUP_TRACK_DEV';
	const browserIdCookieName: string = isProd
		? 'MEETUP_BROWSER_ID'
		: 'MEETUP_BROWSER_ID_DEV';
	const memberCookieName: string = isProd
		? 'MEETUP_MEMBER'
		: 'MEETUP_MEMBER_DEV';

	const trackers: { [string]: Tracker } = getTrackers({
		agent,
		browserIdCookieName,
		memberCookieName,
		trackIdCookieName,
	});

	Object.keys(trackers).forEach((trackType: string) => {
		server.decorate('request', trackType, trackers[trackType], {
			apply: true, // make the `request` available to tracker
		});
	});

	server.ext('onRequest', onRequest);
	server.ext(
		'onPreResponse',
		getOnPreResponse({
			browserIdCookieName,
			memberCookieName,
			trackIdCookieName,
		})
	);

	next();
}

register.attributes = {
	name: ACTIVITY_PLUGIN_NAME,
	version: '1.0.0',
};
