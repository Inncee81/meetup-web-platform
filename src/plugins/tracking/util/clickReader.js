import avro from './avro';
import { parseMemberCookie } from '../../../util/cookieUtils'; // TODO: provide this info through new plugin

const isProd = process.env.NODE_ENV === 'production';
export const clickCookieOptions = {
	isSecure: isProd,
	isHttpOnly: false,
	domain: `${isProd ? '' : '.dev'}.meetup.com`,
};

export const clickToClickRecord = request => click => {
	return {
		timestamp: click.timestamp || new Date().toISOString(),
		requestId: request.id,
		memberId: parseMemberCookie(request.state).id,
		lineage: click.lineage,
		linkText: click.linkText || '',
		coordX: click.coords[0],
		coordY: click.coords[1],
	};
};

export default function processClickTracking(request, reply) {
	const rawCookieValue = (request.state || {})['click-track'];
	// It's possible that multiple cookies with the same value were sent, e.g.
	// one value for .dev.meetup.com and another for .meetup.com - parse only the first
	const cookieValue =
		rawCookieValue instanceof Array ? rawCookieValue[0] : rawCookieValue;
	if (!cookieValue || cookieValue === 'undefined') {
		return;
	}

	const cookieJSON = decodeURIComponent(cookieValue);
	const { history } = JSON.parse(cookieJSON);
	history.map(clickToClickRecord(request)).forEach(avro.loggers.click);

	reply.unstate('click-track', clickCookieOptions);
	return;
}
