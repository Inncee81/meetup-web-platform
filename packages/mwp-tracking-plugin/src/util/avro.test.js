import avsc from 'avsc';
import { getLogger } from '../activity';
import { clickToClickRecord } from './clickReader';
import * as avro from './avro';

jest.mock('@google-cloud/pubsub', () => {
	const publish = jest.fn(() => Promise.resolve('okay!'));
	const main = jest.fn(() => ({
		topic: jest.fn(() => ({
			publisher: () => ({
				publish,
			}),
		})),
	}));
	main.publish = publish; // hook to make publish calls visible to calling test
	return main;
});

describe('serializers.avro', () => {
	it('encodes record of provided schema', () => {
		const schema = {
			type: 'record',
			fields: [
				{ name: 'requestId', type: 'string' },
				{ name: 'timestamp', type: 'string' },
			],
		};
		const serializer = avro.serializers.avro(schema);
		const data = {
			requestId: 'foo',
			timestamp: new Date().getTime().toString(),
		};
		const serialized = serializer(data);

		// parse stringified object
		const valObj = JSON.parse(serialized);
		// create a new buffer from that string
		const avroBuffer = new Buffer(valObj.record, 'base64');
		// get the avro-encoded record
		const recordedInfo = avsc.parse(schema).fromBuffer(avroBuffer);
		expect(recordedInfo).toEqual(data);
	});
});

describe('deserializers.avro', () => {
	it('decodes an encoded record of provided schema', () => {
		const schema = {
			type: 'record',
			fields: [
				{ name: 'requestId', type: 'string' },
				{ name: 'timestamp', type: 'string' },
			],
		};
		const data = {
			requestId: 'foo',
			timestamp: new Date().getTime().toString(),
		};

		const serializer = avro.serializers.avro(schema);
		const deserializer = avro.deserializers.avro(schema);
		const serialized = serializer(data);
		const deserialized = deserializer(serialized);
		expect(deserialized).toEqual(data);
	});
});

describe('Activity tracking', () => {
	const request = {
		id: 'foo',
		headers: {},
		log() {},
	};
	const trackInfo = getLogger('WEB')(request, {
		memberId: 1234,
		trackId: 'foo',
		sessionId: 'bar', // not part of v3 spec
		url: 'asdf',
	});

	it('encodes standard output from getLogger', () => {
		const serialized = avro.serializers.activity(trackInfo);

		// parse stringified object
		const valObj = JSON.parse(serialized);
		// create a new buffer from that string
		const avroBuffer = new Buffer(valObj.record, 'base64');
		// get the avro-encoded record
		const recordedInfo = avsc.parse(avro.schemas.activity).fromBuffer(avroBuffer);
		const expectedTrackedInfo = {
			...trackInfo,
			aggregratedUrl: '', // misspelled, unused field in v3 spec, default ''
			browserId: '',
		};
		delete expectedTrackedInfo.sessionId; // not part of v3 spec
		expect(recordedInfo).toEqual(expectedTrackedInfo);
	});
});

describe('Click tracking', () => {
	const request = {
		id: 'foo',
		state: {},
	};
	const click = {
		timestamp: new Date(0).toISOString(),
		lineage: 'div#foo',
		linkText: 'hello world',
		coords: [23, 45],
	};

	it('encodes standard output from clickToClickRecord', () => {
		const trackInfo = clickToClickRecord(request)(click);
		const serialized = avro.serializers.click(trackInfo);

		// parse stringified object
		const valObj = JSON.parse(serialized);
		// create a new buffer from that string
		const avroBuffer = new Buffer(valObj.record, 'base64');
		// get the avro-encoded record
		const recordedInfo = avsc.parse(avro.schemas.click).fromBuffer(avroBuffer);
		const expectedTrackedInfo = {
			...trackInfo,
			tag: '', // not used in our click data - defaults to empty string
		};
		expect(recordedInfo).toEqual(expectedTrackedInfo);
	});
});

describe('getPlatformAnalyticsLog', () => {
	it('logs to stdout by default', () => {
		spyOn(process.stdout, 'write').and.callThrough();
		const analyticsLog = avro.getPlatformAnalyticsLog();
		analyticsLog('foo');
		expect(process.stdout.write).toHaveBeenCalled();
	});
	it('calls pub/sub topic.publish when isGAE', () => {
		const isGAE = true;
		const analyticsLog = avro.getPlatformAnalyticsLog(isGAE);
		analyticsLog('foo');
		expect(require('@google-cloud/pubsub').publish).toHaveBeenCalled();
	});
});
