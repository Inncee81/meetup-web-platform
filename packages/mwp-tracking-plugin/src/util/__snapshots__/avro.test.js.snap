// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity tracking encodes standard output from getLogger 1`] = `
Activity {
  "agent": "",
  "aggregratedUrl": "",
  "apiVersion": null,
  "browserId": "",
  "ip": "",
  "isUserActivity": true,
  "memberId": 1234,
  "mobileWeb": false,
  "oauthConsumerId": null,
  "parentRequestId": null,
  "platform": "WEB",
  "platformAgent": "WEB",
  "referer": "",
  "requestId": "foo",
  "subViewName": null,
  "timestamp": "2018-05-08T19:50:00.000Z",
  "trackId": "foo",
  "trax": Object {},
  "url": "asdf",
  "viewName": null,
}
`;
